YatimaStdLib/GaloisField.lean

```diff
@@ -0,0 +1,206 @@
+import YatimaStdLib.Polynomial
+import YatimaStdLib.Zmod
+import YatimaStdLib.Bit
+
+/-!
+# Galois Fields
+
+This module provides the basic data structures necessary to define and work with prime fields
+and their extensions.
+
+Here we post some definitions from https://hackage.haskell.org/package/galois-field-1.0.2
+-/
+
+/-- The structure of a Galois field on t-/
+class GaloisField (K : Type _) where
+  plus : K → K → K
+  times : K → K → K
+  null : K
+  ein : K
+  minus : K → K → K
+  divis : K → K → K
+  -- Characteristic `p` of field and order of prime subfield.
+  char : Nat
+  -- Degree `q` of field as extension field over prime subfield.
+  deg : Nat
+  -- Frobenius endomorphism `x ↦ x^p` of prime subfield.
+  frob : K → K
+
+namespace GaloisField
+
+instance [GaloisField K] : Inhabited K where
+  default := null
+
+instance [GaloisField K] : Add K where
+  add := plus
+
+instance [GaloisField K] : Mul K where
+  mul := times
+
+instance [GaloisField K] : OfNat K (nat_lit 1) where
+  ofNat := ein
+
+instance [GaloisField K] : OfNat K (nat_lit 0) where
+  ofNat := null
+
+instance [GaloisField K] : Div K where
+  div := divis
+
+instance [GaloisField K] : Sub K where
+  sub := minus
+
+/-- An `O(log n)` implementation of `galPow` -/
+def fastPow [Mul K] [OfNat K (nat_lit 1)] (x : K) (n : Nat) : K := 
+  let binExp := n.toBits
```

Comment:
Why don't we replace all those constraints with `[GaloisField K]`?

Reply:
I actually need `fastPow` in the instance declaration of `GaloisField` for `Extension K P` (to define the Frobenius), so unfortunately the way I want to use it I don't get to use the `GaloisField K` instance yet :(

Reply:
OK, let's leave it as it is